polkadot/xcm/xcm-builder/src/controller.rs

```diff
@@ -68,12 +72,35 @@ pub trait ExecuteController<Origin, RuntimeCall> {
 		message: Box<VersionedXcm<RuntimeCall>>,
 		max_weight: Weight,
 	) -> Result<Weight, DispatchErrorWithPostInfo>;
+
+	/// Attempt to execute an XCM locally, returns Ok with the weight consumed if the execution
+	/// complete successfully, Err otherwise.
+	///
+	/// # Parameters
+	///
+	/// - `origin`: the origin of the call.
+	/// - `message`: the XCM program to be executed, encoded.
+	/// - `max_weight`: the maximum weight that can be consumed by the execution.
+	fn execute_blob(
+		origin: Origin,
+		message: BoundedVec<u8, MaxXcmEncodedSize>,
+		max_weight: Weight,
+	) -> Result<Weight, DispatchErrorWithPostInfo>;
+
+	/// The underlying XCM execution logic, shared by both `execute` and `execute_blob`.
+	fn execute_base(
+		origin_location: Location,
+		message: Box<VersionedXcm<RuntimeCall>>,
+		max_weight: Weight,
+	) -> Result<Weight, DispatchErrorWithPostInfo>;
```

Comment:
this should probably not belong to the trait, just let it be an implementation detail of whatever implement this